app/controllers/journeys/additional_payments_for_teaching/reminders_form_callbacks.rb

```diff
@@ -0,0 +1,95 @@
+module Journeys
+  module AdditionalPaymentsForTeaching
+    module RemindersFormCallbacks
+      def personal_details_before_show
+        try_mailer { set_a_reminder_immediately_if_possible }
+      end
+
+      def email_verification_before_show
+        try_mailer { set_a_reminder_immediately_if_possible }
+      end
+
+      def email_verification_before_update
+        inject_sent_one_time_password_at_into_the_form
+      end
+
+      def personal_details_after_form_save_success
+        update_reminder_id
+        try_mailer { send_verification_email } || return
```

Comment:
What are your thoughts on sending the verification email from inside the personal details form, or the other forms? Seems like changing email address and sending a verification email are pretty closely related concepts additionally it would let us move some of the form error setting responsibilities out of this module

Reply:
> What are your thoughts on sending the verification email from inside the personal details form, or the other forms? Seems like changing email address and sending a verification email are pretty closely related concepts additionally it would let us move some of the form error setting responsibilities out of this module

@rjlynch, it all boils down to where we decide to put the boundaries of what a form object can and cannot do, which is what prompted the creation of this ticket. I had originally figured that our forms were, in practice, becoming service objects, and I felt ok with that until we decided that it was not ideal and something else had to be done – but as a team, we accepted the fact we were going to complete the form objects epic without a better separation of concerns.

With that context in mind, I think it's best to handle the mailer firing task outside the form object, possibly in controller callbacks or service objects. By separating concerns in this way, we allow the form object to focus on its primary responsibilities (validation, data manipulation and communicating response with the controller), and we avoid violating the single responsibility principle. I understand it's likely we'll always need to verify an email address after changing it, but what should keep those two concepts together is a service object, and in the absence of that, a controller callback – exactly in the same way you wouldn't want to keep those two things together in a model.

Reply:
> I think it's best to handle the mailer firing task outside the form object,

Agreed, the usual approach would be using a job for things like sending emails etc something like

```
def update
  if @some_form.save
   SomeJob.perform_later(@some_form.record)
   redirect_to @some_form.record
  else
    render "edit"
  end
end
```

A bit tricker in our set up but we could infer the job to call to handle post form submission tasks from the name of the slug. Also would probably want to use `perform_now`. Might have the advantage of being easier to test compared to a controller callback?

Reply:
> we could infer the job to call to handle post form submission tasks from the name of the slug

My concern with that is that it's a solution for one specific context – sending emails – while we have seen that there are other generic tasks that need to be performed after a form is saved. So the idea of having a callback is that you can put anything you need in it, and hopefully survives future changes of requirements. That said, I'm open to exploring other solutions if they can propose a re-usable and slug-agnostic pattern.

> Might have the advantage of being easier to test compared to a controller callback?

I take the point about testing, and I didn't even make an attempt at unit testing as those behaviours should be captured in request specs (which we should make more use of, in my opinion) or features specs.

> Probably unrelated to the changes in this pr though, maybe one to discuss with the team

This is a good discussion, and happy to extend it to the rest of the team. I'll try to draft how this applies to the claims controller so we can see what it means at scale.

Reply:
My view on this is that the form object's responsibility should end at validating user input; the persistence of data and other things such as sending e-mails or scheduling jobs should be handled within a service object. This allows us to execute these processes outside of the context of a web form. I think this is the direction we should iterate towards.